### PR TITLE
enable native profile on composite environments bootstrap enabled

### PR DESCRIPTION
**Pull request by aveiros**
Previously we couldn't use native profile on composite environments on SCC 1.4.x

### CODE REVIEW

**Comment by aveiros** [comment]
Please hold on this, with `bootstrap: true` it's suffering from a bug raised against 2.x #975 

I will try to update this PR as soon as i can

**Comment by spencergibb** [comment]
As it is marked for team discussion it won't get merged until we chat about it first.

**Comment by spencergibb** [comment]
Does this work with the `composite` profile?

**Comment by aveiros** [comment]
I'm working on a separate branch, but i managed to get it working with composite profile (using git/native/vault for example).

I need to catch up on tests and that's why haven't updated this PR yet.

I can update this PR (with tests), if you guys want to have a look?

**Comment by aveiros** [comment]
Update on this PR for the 1.4.x branch.

1. Previously we could not use Native Profile with other profiles (i.e. Git, Vault, etc), attachieving this way `composite` Environments. This is now possible using the following configuration:

`spring:
  cloud:
    config:
      server:
        vault:
          host: localhost
          port: 8200
          backend: test
          defaultKey: test
        git:
          uri: file://${HOME}/config
        native:
          searchLocations: file://${HOME}/config-native`

**Comment by spencergibb** [comment]
Was this only for if `bootstrap=true`?

**Comment by spencergibb** [comment]
After looking at this, I found I could get the desired behaviour by adding `CompositeConfiguration.class` after `EnvironmentRepositoryConfiguration.class` here https://github.com/spring-cloud/spring-cloud-config/blob/5340f2d77ae8fe13535b627d86a5813b410caff5/spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/bootstrap/ConfigServerBootstrapConfiguration.java#L48

**Comment by aveiros** [comment]
> After looking at this, I found I could get the desired behaviour by adding `CompositeConfiguration.class` after `EnvironmentRepositoryConfiguration.class` here
> 
> [spring-cloud-config/spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/bootstrap/ConfigServerBootstrapConfiguration.java](https://github.com/spring-cloud/spring-cloud-config/blob/5340f2d77ae8fe13535b627d86a5813b410caff5/spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/bootstrap/ConfigServerBootstrapConfiguration.java#L48)
> 
> Line 48 in [5340f2d](/spring-cloud/spring-cloud-config/commit/5340f2d77ae8fe13535b627d86a5813b410caff5)
> 
>  @Import({ EnvironmentRepositoryConfiguration.class, TransportConfiguration.class }) 
> Are you willing to rework your PR?

Sure!

This PR includes 2 changes
1) Native Environment Repository can now be used as a "Composite Environment Repository". Previously you could not do this on 1.4.x.

Previously you could only combine git,vault or svn,vault or git,svn. Now you can also include native along with those ones.

Docs => https://cloud.spring.io/spring-cloud-static/spring-cloud-config/1.4.5.RELEASE/single/spring-cloud-config.html#_composite_environment_repositories


2) Fix for #975, where when using `bootstrap: true`, SCC would just fall over.

**Comment by spencergibb** [comment]
I was unable to reproduce number 1. It worked fine for me when not using bootstrap.

**Comment by aveiros** [comment]
sorry

native+git => works fine
native+git+vault => native never gets instantiated

server config:
`spring:
  application:
    name: configserver
  profiles:
    active: native,git,vault
  cloud:
    config:
      server:
        vault:
          host: localhost
          port: 8200
          backend: myapp
          order: 2
        git:
          uri: file://${HOME}/git
          order: 1
        native:
          searchLocations: file://${HOME}/native
          order: 0
`

**Comment by aveiros** [comment]
tested with:
native,git => works
native,git,vault => works
native,vault => works

worth putting a Vault Int Test?

File: spring-cloud-config-sample/src/test/java/sample/ApplicationBootstrapTests.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
+package sample;
+
```
[the remark this answers is not in the document]
Comment (aveiros) [reply spring-cloud-config-sample/src/test/java/sample/ApplicationBootstrapTests.java]: done

File: spring-cloud-config-sample/src/main/resources/bootstrapservercomposite.yml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+spring:
```
[the remark this answers is not in the document]
Comment (aveiros) [reply spring-cloud-config-sample/src/main/resources/bootstrapservercomposite.yml]: good spot, master is suffering from the same issue

File: spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/VaultEnvironmentRepository.java [hunk 3]
```diff
@@ -86,11 +87,11 @@
 	private RestTemplate rest;
 
 	//TODO: move to watchState:String on findOne?
-	private HttpServletRequest request;
+	private ObjectProvider<HttpServletRequest> request;
 
 	private EnvironmentWatch watch;
 
-	public VaultEnvironmentRepository(HttpServletRequest request, EnvironmentWatch watch, RestTemplate rest) {
+	public VaultEnvironmentRepository(ObjectProvider<HttpServletRequest> request, EnvironmentWatch watch, RestTemplate rest) {
```
[the remark this answers is not in the document]
Comment (aveiros) [reply spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/VaultEnvironmentRepository.java]: Ported code from Spencer from master/2.x (links below). Not clear how could I get it working with HttpServletRequest or ObjectProvider<HttpServletRequest>... 


Ticket =>  https://github.com/spring-cloud/spring-cloud-config/issues/975
PR => https://github.com/spring-cloud/spring-cloud-config/commit/2b1248da48dbb0011a206a8805617b2decc40508

**Comment by spencergibb** [comment]
Closing this due to inactivity. Please re-open if there's more to discuss.

**Pull request closed without merging**